### PR TITLE
feat(quantic): query pipeline trigger logic added to Quantic Did You Mean component

### PR DESCRIPTION
**Pull request by mmitiche**
[SFINT-4881](https://coveord.atlassian.net/browse/SFINT-4881)

In this PR:

- Added the query pipeline trigger logic added to Quantic Did You Mean component.
- Added Cypress E2E tests to cover the logic of the Quantic Did You Mean component.

### Demo:

Rule

https://github.com/coveo/ui-kit/assets/86681870/9dc4a7e4-dbf3-4517-a9ae-447277280404

[SFINT-4881]: https://coveord.atlassian.net/browse/SFINT-4881?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml [hunk 1]
```diff
@@ -1260,4 +1260,18 @@
...
+      <protected>false</protected>
+      <shortDescription>Showing results for example</shortDescription>
+  </labels>
+  <labels>
+    <fullName>quantic_SearchInsteadFor</fullName>
+    <value>Search instead for</value>
+    <language>en_US</language>
+    <protected>false</protected>
+    <shortDescription>Search instead for</shortDescription>
+  </labels>
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:1276]: Why did we need this now?

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.html
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.html:1]: There's a way to specify that the same file was renamed I think to git when doing your commit, but it's not a huge problem. And I also forget to do it most of the time

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js [hunk 2]
```diff
@@ -39,18 +49,24 @@ export default class QuanticDidYouMean extends LightningElement {
   @track correctedQuery;
 
   /** @type {Function} */
-  unsubscribe;
+  unsubscribeDidYouMean;
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:52]: Do you also want a `unsubscribeQueryTrigger` ?

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js [hunk 3]
```diff
@@ -67,25 +83,40 @@ export default class QuanticDidYouMean extends LightningElement {
   initialize = (engine) => {
     this.headless = getHeadlessBundle(this.engineId);
     this.didYouMean = this.headless.buildDidYouMean(engine);
-    this.unsubscribe = this.didYouMean.subscribe(() => this.updateState());
+    this.queryTrigger = this.headless?.buildQueryTrigger?.(engine);
+    this.unsubscribeDidYouMean = this.didYouMean.subscribe(() =>
+      this.updateDidYouMeanState()
+    );
+    this.unsubscribeQueryTrigger = this.queryTrigger?.subscribe(() =>
```
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:90]: Why can the `this.queryTrigger.subscribe` be null but not the `this.didYouMean.subscribe`?

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/templates/didYouMean.html
Comment (erocheleau) [inline packages/quantic/force-app/main/default/lwc/quanticDidYouMean/templates/didYouMean.html:1]: No change in essence to the previous didYouMean template right?

File: packages/quantic/cypress/e2e/default-1/did-you-mean/did-you-mean.cypress.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+          Expect.displayApplyCorrectionButton(false);
+          Expect.displayNoResultLabel(true);
+          Expect.noResultLabelContains(exampleOriginalQuery);
+          Expect.displayAutomaticQueryCorrectionLabel(true);
+          Expect.automaticQueryCorrectionLabelContains(exampleCorrectedQuery);
+          Expect.logDidYouMeanAutomaticCorrection();
+        });
+      });
+
+      if (param.useCase === useCaseEnum.search) {
```
Comment (erocheleau) [inline packages/quantic/cypress/e2e/default-1/did-you-mean/did-you-mean.cypress.ts:113]: Does it make sense to split the test in a pipeline trigger spec file instead?
What do you think?

File: packages/quantic/cypress/e2e/default-1/did-you-mean/did-you-mean.cypress.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        describe('query pipeline triggers', () => {
+          describe('when a query pipeline trigger is fired', () => {
```
Comment (erocheleau) [inline packages/quantic/cypress/e2e/default-1/did-you-mean/did-you-mean.cypress.ts:115]: Any specific corner case you can imagine for example a query pipeline trigger that would change the query to show no results, or break or not work, something like that that would be missing from those tests?

**Review by erocheleau — APPROVED** [review]
Looks good to me, good work!
A few questions

File: packages/quantic/force-app/main/translations/fr.translation-meta.xml [hunk 6]
```diff
@@ -691,5 +691,13 @@
...
     <label>Quelque chose s'est mal passé lors de la tentative de génération d'une réponse.</label>
     <name>quantic_CouldNotGenerateAnAnswer</name>
-</customLabels>
+  </customLabels>
+  <customLabels>
+    <label>Résultats affichés pour {{0}}</label>
+    <name>quantic_ShowingResultsFor</name>
+  </customLabels>
+  <customLabels>
+    <label>Rechercher à la place {{0}}</label>
```
Comment (SimonMilord) [inline packages/quantic/force-app/main/translations/fr.translation-meta.xml:700]:
```suggestion
    <label>Rechercher plutôt {{0}}</label>
```

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js [hunk 7]
```diff
@@ -100,10 +131,39 @@ export default class QuanticDidYouMean extends LightningElement {
     );
   }
 
+  get showingResultsForLabel() {
+    return I18nUtils.format(
+      this.labels.showingResultsFor,
+      I18nUtils.getTextBold(I18nUtils.escapeHTML(this.newTriggeredQuery))
+    );
+  }
+
```
Comment (SimonMilord) [inline packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:140]: should we leave a comment here to say that this is for e2e tests purposes?

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js (hunk 3 above)
Comment (SimonMilord) [reply packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:90]: yeah not sure I understand that part

**Review by SimonMilord — APPROVED** [review]
Nice job!

File: packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml (hunk 1 above)
Comment (mmitiche) [reply packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:1276]: These are specific labels that we use when the Query is updated by a pipeline trigger, when instead the query is corrected because it contained a typo or something, other labels will be used/. for example: `Query was automatically corrected to "example correct query"

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js (hunk 3 above)
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:90]: Because the QueryTrigger is not yet available in the Insight Use Case, currently the query trigger Headless controller only in the search use case, while the Headless did you mean controller is available in both use cases.

that's why I'm adding these checks, to avoid breaking the Quantic Did You Mean component when being used in the insight use case.

We will need to update Headless in the futur tho, to make it support the triggers for the insight use case.

File: packages/quantic/cypress/e2e/default-1/did-you-mean/did-you-mean.cypress.ts (hunk 4 above)
Comment (mmitiche) [reply packages/quantic/cypress/e2e/default-1/did-you-mean/did-you-mean.cypress.ts:113]: Sure yeah this makes sense 👌

File: packages/quantic/cypress/e2e/default-1/did-you-mean/did-you-mean.cypress.ts (hunk 5 above)
Comment (mmitiche) [reply packages/quantic/cypress/e2e/default-1/did-you-mean/did-you-mean.cypress.ts:115]: very good point I ll try out some corner cases and see if any require some  new tests.

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js (hunk 7 above)
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticDidYouMean/quanticDidYouMean.js:140]: This is not for e2e tests purposes, what is making you think that it is?
This just for showing the right label with the query in a bold format.

File: packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml (hunk 1 above)
Comment (mmitiche) [inline packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:1276]: Hello @cmartel-coveo ! These are the labels we display in Quantic when the query is updated by a pipeline trigger:

labels

Please let me know if you think anything we can improve with these labels! Thank you!

File: packages/quantic/force-app/main/default/lwc/quanticDidYouMean/templates/didYouMean.html
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticDidYouMean/templates/didYouMean.html:1]: Exactly 👌

File: packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml (hunk 1 above)
Comment (cmartel-coveo) [reply packages/quantic/force-app/main/default/labels/CustomLabels.labels-meta.xml:1276]: @mmitiche Sorry for the late reply... But your copy is good, so no worries :)
